src/haz3lcore/dynamics/TypeAssignment.re

```diff
@@ -0,0 +1,288 @@
+open Util;
+open OptUtil.Syntax;
+
+let equal_typ_case = (l: list(Typ.t)): option(Typ.t) => {
+  switch (l) {
+  | [] => None
+  | _ =>
+    let ty = List.hd(l);
+    List.fold_left((acc, t) => {acc && Typ.eq(t, ty)}, true, l)
+      ? Some(ty) : None;
+  };
+};
+
+let rule_prj = (dr: DHExp.rule): (DHPat.t, DHExp.t) => {
+  switch (dr) {
+  | Rule(dhp, dh) => (dhp, dh)
+  };
+};
+
+let arrow_aux = (ty: Typ.t): Typ.t => {
+  switch (ty) {
+  | Unknown(Internal) => Arrow(Unknown(Internal), Unknown(Internal))
+  | _ => ty
+  };
+};
+
+let delta_ty = (id: MetaVar.t, m: Statics.Map.t): option(Typ.t) => {
+  switch (Id.Map.find_opt(id, m)) {
+  | Some(InfoExp({mode, ctx, _})) =>
+    switch (mode) {
+    | Syn
+    | SynFun => Some(Unknown(Internal))
+    | Ana(ana_ty) => Some(Typ.normalize(ctx, ana_ty))
+    }
+  | _ => None
+  };
+};
+
+let ground = (ty: Typ.t): bool => {
+  switch (ty) {
+  | Bool
+  | Int
+  | Float
+  | String
+  | Prod([])
+  | Arrow(Unknown(Internal), Unknown(Internal)) => true
+  | _ => false
+  };
+};
+
+let rec dhpat_extend_ctx = (dhpat: DHPat.t, ty: Typ.t, ctx: Ctx.t): Ctx.t => {
+  switch (dhpat, ty) {
+  | (Var(name), _) =>
+    let entry = Ctx.VarEntry({name, id: Id.invalid, typ: ty});
+    Ctx.extend(ctx, entry);
+  | (Tuple(l1), Prod(l2)) =>
+    if (List.length(l1) == List.length(l2)) {
+      List.fold_left2(
+        (acc, dhp, typ) => {dhpat_extend_ctx(dhp, typ, acc)},
+        ctx,
+        l1,
+        l2,
+      );
+    } else {
+      ctx;
+    }
+  | (Cons(dhp1, dhp2), List(typ)) =>
+    ctx |> dhpat_extend_ctx(dhp1, typ) |> dhpat_extend_ctx(dhp2, ty)
+  | (ListLit(typ1, l), List(typ2)) =>
+    if (Typ.eq(typ1, typ2)) {
+      List.fold_left(
+        (acc, dhp) => {dhpat_extend_ctx(dhp, typ1, acc)},
+        ctx,
+        l,
+      );
+    } else {
+      ctx;
+    }
+  | (Ap(Constructor(_, typ), dhp), _) =>
+    let (ty1, ty2) = Typ.matched_arrow(ctx, typ);
+    if (Typ.eq(ty2, ty)) {
+      ctx |> dhpat_extend_ctx(dhp, ty1);
+    } else {
+      ctx;
+    };
+  | _ => ctx
+  };
+};
+
+let rec typ_of_dhexp =
+        (ctx: Ctx.t, m: Statics.Map.t, dh: DHExp.t): option(Typ.t) => {
+  switch (dh) {
+  | EmptyHole(id, _) => delta_ty(id, m)
+  | NonEmptyHole(_, id, _, d) =>
+    switch (typ_of_dhexp(ctx, m, d)) {
+    | None => None
+    | Some(_) => delta_ty(id, m)
+    }
+  | ExpandingKeyword(_) => None
+  | FreeVar(_) => Some(Unknown(Internal))
+  | InvalidText(_) => None
+  | InconsistentBranches(_, _, Case(d_scrut, d_rules, _)) =>
+    let* ty' = typ_of_dhexp(ctx, m, d_scrut);
+    let typ_cases =
+      d_rules
+      |> List.map(rule_prj)
```

Comment:
you don't need to do this map -- you can directly pattern match in the argument position on the next line